samples/c/create_systemd_scope.c

```diff
@@ -0,0 +1,614 @@
+		goto error;
+	}
+
+	ret = cgroup_delete_cgroup(cg, 1);
+	if (ret)
+		printf("Failed to delete the %s cgroup: %d: %s\n", TMP_CGNAME, ret,
+		       cgroup_strerror(ret));
+
+error:
+	if (pids)
+		free(pids);
+
+	if (cg)
+		cgroup_free(&cg);
+
+	if (ret == 0 && saved_errno)
+		ret = ECGFAIL;
+
+	return ret;
+}
+
+static void wait_for_child(pid_t pid)
+{
+	int wstatus;
+
+	(void)waitpid(pid, &wstatus, 0);
+
+	if (WIFEXITED(wstatus))
+		printf("pid %d exited with status %d\n", pid, WEXITSTATUS(wstatus));
+	else if (WIFSIGNALED(wstatus))
+		printf("pid %d exited due to signal %d\n", pid, WTERMSIG(wstatus));
+}
+
+int main(int argc, char *argv[])
+{
+	pid_t high_pid, med_pid, low_pid;
+	struct example_opts opts = {0};
+	int ret = 0;
+
+	if (argc < 3) {
+		usage(argv[0]);
+		exit(1);
+	}
+
+	ret = cgroup_set_default_scope_opts(&opts.systemd_opts);
+	if (ret) {
+		printf("Failed to set the default systemd options: %d\n", ret);
+		goto error;
+	}
+
+	ret = parse_opts(argc, argv, &opts);
+	if (ret) {
+		printf("Failed to parse the command line options: %d\n", ret);
+		goto error;
+	}
+
+	if (opts.debug)
+		cgroup_set_default_logger(CGROUP_LOG_DEBUG);
+
+	ret = cgroup_init();
+	if (ret)
+		goto error;
+
+	ret = create_scope(&opts);
+	if (ret)
+		goto error;
+
+	ret = create_tmp_cgroup(&opts);
+	if (ret)
+		goto error;
+
+	ret = move_pids_to_tmp_cgroup(&opts);
+	if (ret)
+		goto error;
+
+	ret = create_high_priority_cgroup(&opts);
+	if (ret)
+		goto error;
+
+	ret = create_medium_priority_cgroup(&opts);
+	if (ret)
+		goto error;
+
+	ret = create_low_priority_cgroup(&opts);
+	if (ret)
+		goto error;
+
+	ret = create_process(&high_pid, HIGH_CGNAME);
+	if (ret)
+		goto error;
+
+	ret = create_process(&med_pid, MED_CGNAME);
+	if (ret)
+		goto error;
+
+	ret = create_process(&low_pid, LOW_CGNAME);
+	if (ret)
+		goto error;
+
+	ret = delete_tmp_cgroup();
+	if (ret)
+		goto error;
+
+	printf("\n----------------------------------------------------------------\n");
+	printf("Cgroup setup completed successfully\n");
+	printf("\t* The scope %s was placed under slice %s\n", opts.scope, opts.slice);
+	printf("\t* Libcgroup initially placed an idle process in the scope,\n"
+	       "\t  but it has been removed by this program\n");
+	printf("\t* PID %d has been placed in the %s cgroup\n", high_pid, HIGH_CGNAME);
+	printf("\t* PID %d has been placed in the %s cgroup\n", med_pid, MED_CGNAME);
+	printf("\t* PID %d has been placed in the %s cgroup\n", low_pid, LOW_CGNAME);
+	printf("\nThis program will wait for the aforementioned child processes to\n"
+	       "exit before exiting itself. Systemd will automatically delete the\n"
+	       "scope when there are no longer any processes running within the.\n"
+	       "scope.  Systemd will not automatically delete the slice.\n");
+	printf("----------------------------------------------------------------\n\n");
+
+	wait_for_child(high_pid);
+	wait_for_child(med_pid);
+	wait_for_child(low_pid);
+
+error:
+	return ret;
+}
```

Comment:
@drakenclimber This is a pretty cool training manual, instead of a sample. Thank you so much for making it so extensive.